docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,175 @@
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/third-party-api-servers.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Third Party API Servers
+
+## Goal
+
+With third party API servers, we want to enable cluster admins or users to
+expose new APIs at runtime without having to modify the core kubernetes API
+server binary.
+Anyone can write their own API server and register it with the core
+kubernetes API server. These new APIs will be seamless extension of core
+kubernetes APIs (ex: they will be operated upon via kubectl).
+
+## Non Goals:
+
+The following are related but are not the goals of this specific proposal:
+* Make it easy to write a kubernetes API server.
+
+## High Level Architecture
+
+Users can bring up their own API servers and register them with the core server
+at runtime.
+Core server will then start exposing these third party servers in the discovery
+API. kubectl will use the discovery API to connect to these servers.
+
+These API servers will have full control of their resources.
+Each server will have their own storage mechanism and will implement their own
+authorization and authentication.
+
+
+## Implementation Details
+
+### Thirdpartyapiservers Extensions Resource
+
+API server will support a non-namespaced `thirdpartyapiservers` extension resource.
+Third party servers can POST at
+`http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to register themselves.
+Clients can GET `http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to get a list of registered third party servers.
+The API definition will be as follows:
+
+```
+// ThirdPartyAPIServer represents the configuration of a third party server.
+type ThirdPartyAPIServer struct {
+  unversioned.TypeMeta
+
+  ObjectMeta
+
+  Spec ThirdPartyAPIServerSpec
+
+  Status ThirdPartyAPIServerStatus
+}
+
+// ThirdPartyAPIServerSpec is the specification of a ThirdPartyAPIServer
+// resource.
+type ThirdPartyAPIServerSpec struct {
+  // Base path of the API server.
+  BasePath string
+
+  // Path to the swagger spec.
+  SwaggerSpecPath string
+}
+
+type ThirdPartyAPIServerStatus struct {
+  Conditions []ThirdPartyAPIServerCondition
+}
+
+type ThirdPartyAPIServerConditionType string
+const (
+        ThirdPartyAPIServerValid ThirdPartyAPIServerConditionType = "Valid"
+)
+
+type ThirdPartyAPIServerCondition struct {
+  Type ThirdPartyAPIServerConditionType
+  // Status of the condition, one of True, False or Unknown.
+  Status ConditionStatus
+  Reason string
+  Message string
+}
+```
+
+There are a few requirements that the third party API servers should meet
+(for ex: name of resources provided by each server should be unique).
```

Comment:
No, the unit that they export is API group. API groups are domain names (mygroup.mycompany.com). Resource collisions don't matter, since the group disambiguates.